src/common/bitset_set.h

```diff
@@ -0,0 +1,442 @@
+   * modern CPUs.
+   */
+  size_t size() const {
+    size_t count = 0;
+    for (size_t i = 0; i < WORDS; ++i) {
+      count += std::popcount(words[i]);
+    }
+    return count;
+  }
+
+  /** @return maximum size of set  */
+  constexpr size_t max_size() const {
+    return MAX;
+  }
+
+  /** Utility for encode/decode to allow serialisations */
+  void bound_encode(size_t &p) const {
+    for (size_t i = 0; i < WORDS; ++i) {
+      denc_varint(words[i], p);
+    }
+  }
+
+  /** Utility for encode/decode to allow serialisations */
+  void encode(ceph::buffer::list::contiguous_appender &bl) const {
+    for (size_t i = 0; i < WORDS; ++i) {
+      denc_varint(words[i], bl);
+    }
+  }
+
+  /** Utility for encode/decode to allow serialisations */
+  void decode(ceph::buffer::ptr::const_iterator &bp) {
+    for (size_t i = 0; i < WORDS; ++i) {
+      denc_varint(words[i], bp);
+    }
+  }
+
+  /** @return begin const_iterator. There is no non-const iterator */
+  const_iterator begin() const {
+    return const_iterator(this);
+  }
+
+  /** @return begin const_iterator. There is no non-const iterator */
+  const_iterator cbegin() const {
+    return begin();
+  }
+
+  /** @return begin const_iterator. There is no non-const iterator */
+  const_iterator end() const {
+    return _end;
+  }
+
+  /** @return begin const_iterator. There is no non-const iterator */
+  const_iterator cend() const {
+    return _end;
+  }
+
+  /** Utility for implemting copy operators.  Not intented to be used
+   * externally, although it is safe to do so
+   */
+  void copy(const bitset_set &other) {
+    for (size_t i = 0; i < WORDS; ++i) {
+      words[i] = other.words[i];
+    }
+  }
+
+  /** Assign-with-copy operator */
+  bitset_set &operator=(const bitset_set &other) {
+    copy(other);
+    return *this;
+  }
+
+  /** Assign with move operator */
+  bitset_set &operator=(bitset_set &&other) noexcept {
+    copy(other);
+    return *this;
+  }
+
+  /** Swap contents with other. */
+  void swap(bitset_set &other) noexcept {
+    for (size_t i = 0; i < WORDS; ++i) {
+      uint64_t tmp = other.words[i];
+      other.words[i] = words[i];
+      words[i] = tmp;
+    }
+  }
+
+  /** Returns true if other contains a subset of the keys in this container.
+   *
+   * Useful to replace std::includes if comparing complete sets
+   */
+  bool includes(const bitset_set &other) const {
+    for (size_t i = 0; i < WORDS; ++i) {
+      if ((words[i] & other.words[i]) != other.words[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /** Standard copy operator */
+  friend bool operator==(const bitset_set &lhs, const bitset_set &rhs) {
+    for (size_t i = 0; i < WORDS; ++i) {
+      if (lhs.words[i] != rhs.words[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /** Standard ostream operator */
+  friend std::ostream &operator<<(std::ostream &lhs, const bitset_set &rhs) {
```

Comment:
Please create an fmt formatter for the bit set (and possibly - implement the operator<< using it)